File: genSNPs.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import csv
import numpy
import random
import time
import math
from csvSNPsAgent import csvSNPsAgent

def generateNewSNPs(genotypes, dis_snp_pos):
    x_spec = 0
    prob = 0
    alpha = 0.0
    gene_effect = 0.0
    disease = 0

    random.seed(time.time())
    sample1 = random.randint(0, len(genotypes)-1)
    sample2 = random.randint(0, len(genotypes)-1)

    while sample1 == sample2:
        sample2 = random.randint(0, len(genotypes)-1)

    x_spec = genotypes[sample1][dis_snp_pos] + genotypes[sample2][dis_snp_pos]

    # Find probabiliy value using logistic regression
    exp_value = math.exp(alpha + (gene_effect * x_spec))
    prob = exp_value / (1 + exp_value)

    # Random one number [0...1] by using uniform random
    runif = numpy.random.uniform(0,1,1)

    if runif[0] < prob:
        disease = 1
    else:
        disease = 0

    # Create new genotypes form original data (a + b)
    new_genotypes = [x + y for x, y in zip(genotypes[sample1], genotypes[sample2])]

    return disease, new_genotypes

def main():
    outfile = '../result/'+str(time.time())+'.csv'
    individual_size = 3008
    number_of_snp = 13479
    dis_snp = 'rs3789038'

    number_of_population = 1000
    number_of_case = number_of_population / 2
    all_data = numpy.zeros((number_of_population, number_of_snp + 1), dtype=numpy.int)

    all_data_index = 0
    case_index = 0
    control_index = number_of_case

    # Read Data form CSV file
    csvAgent = csvSNPsAgent()

    snpsName = csvAgent.readSNPsName()
    genotypes = csvAgent.readSNPsData(individual_size, number_of_snp)

    # Find position of snp name 'rs3789038'
    dis_snp_pos = snpsName.index(dis_snp)

    # Iterative generate 'Case' and 'Control' genotypes
    while case_index < number_of_case or control_index < number_of_population:
        disease = 0
        all_data_index = -1
        new_genotypes = []

        # Generate new genotypes form real data
        disease, new_genotypes = generateNewSNPs(genotypes, dis_snp_pos)

        # Combine Y value (label) and new genotypes
        # By add Y value into index 0 of list
        new_genotypes.insert(0, disease)

        # Classify genotypes between 'Case' and 'Control'
        # Disease = 1 is 'Case'
        # Disease = 0 is 'Control'
        if disease == 1 and case_index < number_of_case:
            all_data_index = case_index
            case_index += 1
        elif disease == 0 and control_index < number_of_population:
            all_data_index = control_index
            control_index += 1

        # Stroe new genotype into all_data
        # AllData[individual][genotypes]
        if all_data_index != -1:
            all_data[all_data_index] = new_genotypes

    # Write all_data to CSV file
    csvAgent.writeAllDataToCSV(outfile, all_data, number_of_snp)

if __name__ == '__main__':
    main()

File: gen-snps.R
# Code by Phisan Sookkhee and Sirikanlaya Sookkhee

# Funciton for generate only one new genotype
# @param genotype : matrix for store genotype data from CSV file
# @param dis_snp_pos : position of SNPs for predict 'Case' or 'Control'
# @return new_genotype
generateGenotype <- function(genotypes, dis_snp_pos) {
     # Generate new genotype from real data
     # @gen: Step 1 - random 2 samples form genotypes
     set.seed(as.numeric(Sys.time()))
     sample_total <- nrow(genotypes)
     samples_index <- sample(1:sample_total, 2, replace = T)

     # @gen: Step 2 - calculate x_spec
     x_spec <- as.numeric(genotypes[samples_index[1], dis_snp_pos]
                          + genotypes[samples_index[2], dis_snp_pos])

     # @gen: Step 3 - Find probabiliy value using logistic regression
     alpha <- 0.0
     gene_effect <- 0.0

     exp_value <- exp(alpha + (gene_effect * x_spec))
     prob <- exp_value / (1 + exp_value)

     # @gen: Step 4 - Using uniform random to cutoff \
     #                new genotype is 'Case' or 'Control'
     cutoff_value <- runif(1, 0, 1)
     if (cutoff_value < prob) {
          disease <- 1
     }else {
          disease <- 0
     }

     # @gen: Step 5 - create new Individual by add vector A and vector B together
     tmp_genotype <-
          genotypes[samples_index[1],] + genotypes[samples_index[2],]
     tmp_genotype <- as.list(tmp_genotype)

     # @gen: step 6 - insert disease or Y value into fist index of list
     new_genotype <- append(tmp_genotype, disease, 0)

     # @gen : step 7 - return only one new genotype
     return(new_genotype)
}

#--------------------------------------------------------------------
# Main Program

# Config values
setwd("~/ResearchCode/SNPsR")
snpsName_path <- "sourceFile/NameONLY13479.csv"
genotype_path <- "sourceFile/New_data_3008persons_13479SNPs.csv"
outfile <- paste("result/", format(Sys.time(), "%b-%d-%Y-%X"), ".csv", sep = "")

dis_snp = 'rs3789038'
number_of_population <- 1000
number_of_case <- number_of_population / 2

# Read input data from CSV files
snpsName_data <- read.csv(snpsName_path, header = TRUE)
genotypes_data <- read.csv(genotype_path, header = TRUE)

# Get number of SNPs from genotype data frame
snps_total <- ncol(genotypes_data)

# Convert snpsName data frame to vector
# And find SNPs position
snpsName <- snpsName_data[,1]
dis_snp_pos <- match(dis_snp, snpsName)

# Convert genotype data frame to matrix
genotypes <- data.matrix(genotypes_data)

# Create matrix for store 'Case' and 'Control'
# 1 - 500 : is a Case
# 501 - 1000 is a Control
all_data <- matrix(0, nrow = number_of_population, ncol = snps_total + 1)
case_index  <- 0
control_index <- number_of_case

while (case_index < number_of_case || control_index < number_of_population) {

     all_data_index <- -1

     # Call function generateGenotype
     # generateGenotype return new genotype as a List
     new_genotype <- generateGenotype(genotypes, dis_snp_pos)

     # Classify genotypes between 'Case' and 'Control' by using Y value
     # y_value = 1 is 'Case'
     # y_value = 0 is 'Control'
     y_value <- new_genotype[[1]]

     if (y_value == 1 && case_index < number_of_case) {
          case_index <- case_index + 1
          all_data_index <- case_index
     } else if (y_value == 0 &&
                control_index < number_of_population) {
          control_index <- control_index + 1
          all_data_index <- control_index
     }

     # Store new gentoype with Y value into all_data
     # and classify it is 'Case' or 'Control'
     if (all_data_index != -1) {
          for (i in 1:(snps_total + 1)) {
               all_data[all_data_index, i] <- new_genotype[[i]]
          }
     }
}

# Set header to first row of result matrix
colnames(all_data) <- c("Y value", paste("SNP", 1:snps_total))

# Write all data to CSV
write.csv(all_data, file = outfile, row.names = FALSE)
